File: work/DSA/c++/recursion_1b/replacePi.cpp
#include <bits/stdc++.h>
using namespace std;

void shiftCharacters(char input[])
{
    for (int i = strlen(input) - 1; i >= 0; i--)
    {
        input[i + 2] = input[i];
    }
}
void replacePi(char input[])
{
    if (input[1] == '\0')
    {
        return;
    }
    replacePi(input + 1);
    if (input[0] == 'p' && input[1] == 'i')
    {
        shiftCharacters(input + 2);

        input[0] = '3';
        input[1] = '.';
        input[2] = '1';
        input[3] = '4';
    }
}

int main()
{
    ios::sync_with_stdio(0);
    cin.tie(0);

    char input[10000];
    cin.getline(input, 10000);
    replacePi(input);
    cout << input << endl;
}

File: work/DSA/javascript/recursion_1b/replacePi.js
let input = ["p", "i", "p", "i"]

function shiftCharacters(input) {
    for (let i = input.length - 1; i >= 0; i--) {
        input[i + 2] = input[i]
    }
}

function replacePi(input, currentIndex) {
    console.log(input[currentIndex])
    if (input[currentIndex + 1] == undefined) {
        return
    }
    replacePi(input, currentIndex + 1)
    if (input[currentIndex] === 'p' && input[currentIndex + 1] === 'i') {
        shiftCharacters(input, currentIndex + 3)
        input[currentIndex] = '3'
        input[currentIndex + 1] = '.'
        input[currentIndex + 2] = '1'
        input[currentIndex + 3] = '4'
    }
}

replacePi(input, 0)
console.log(input)